lta/replicator.py

```diff
@@ -0,0 +1,117 @@
+
+import asyncio
+import json
+from logging import Logger
+import logging
+import sys
+from typing import Any, Dict, Optional
+
+from rest_tools.client import RestClient  # type: ignore
+
+from .component import COMMON_CONFIG, Component, status_loop, work_loop
+from .config import from_environment
+from .log_format import StructuredFormatter
+
+EXPECTED_CONFIG = COMMON_CONFIG.copy()
+EXPECTED_CONFIG.update({
+    "LTA_SITE_CONFIG": "etc/site.json",
+    "WORK_RETRIES": "3",
+    "WORK_TIMEOUT_SECONDS": "30",
+})
+
+
+class Replicator(Component):
+    """
+    Replicator is a Long Term Archive component.
+
+    A Replicator is responsible for registering completed archive bundles
+    with the Rucio transfer service. Rucio will then replicate the bundle
+    from the source (i.e.: WIPAC Data Warehouse) to the destination(s),
+    (i.e.: DESY, NERSC DTN).
+
+    It uses the LTA DB to find completed bundles that need to be registered.
+    It registers the bundles with Rucio. It updates the Bundle and the
+    corresponding TransferRequest in the LTA DB with a 'transferring' status.
+    """
+
+    def __init__(self, config: Dict[str, str], logger: Logger) -> None:
+        """
+        Create a Replicator component.
+
+        config - A dictionary of required configuration values.
+        logger - The object the bundler should use for logging.
+        """
+        super(Replicator, self).__init__("replicator", config, logger)
+        self.work_retries = int(config["WORK_RETRIES"])
+        self.work_timeout_seconds = float(config["WORK_TIMEOUT_SECONDS"])
+        with open(config["LTA_SITE_CONFIG"]) as site_data:
+            self.lta_site_config = json.load(site_data)
+        self.sites = self.lta_site_config["sites"]
+
+    def _do_status(self) -> Dict[str, Any]:
+        """Replicator has no additional status to contribute."""
+        return {}
+
+    def _expected_config(self) -> Dict[str, Optional[str]]:
+        """Replicator provides our expected configuration dictionary."""
+        return EXPECTED_CONFIG
+
+    async def _do_work(self) -> None:
+        """Perform a work cycle for this component."""
+        await self._consume_bundles_to_replicate()
+
+    async def _consume_bundles_to_replicate(self) -> None:
+        """Consume bundles from the LTA DB and register them with Rucio."""
+        # TODO: Perform the work against rucio
+        # 1. Pop bundles from the LTA DB
+        # 2. for each bundle
+        # 2.1. upload and regsiter the bundle to rucio
+        #     rucio upload --rse $RSE --scope SCOPE --register-after-upload --pfn PFN --name NAME /PATH/TO/BUNDLE
+        # 2.2. for each destination site
+        # 2.2.1. add the BUNDLE_DID from 2.1 to the replica
+        #     rucio attach DEST_CONTAINER_DID BUNDLE_DID
+        # 2.3. update the Bundle in the LTA DB; registration information
+        # 2.4. update the TransferRequest in the LTA DB; state: [None] -> [Transferring]
```

Comment:
Use Rucio REST API.

Reply:
The Rucio client is a pretty thick wrapper over the REST API.

I think this is the entry point for the `rucio upload` command:
https://github.com/rucio/rucio/blob/master/bin/rucio#L828

Which goes down through several layers:
https://github.com/rucio/rucio/blob/master/lib/rucio/client/uploadclient.py#L71
https://github.com/rucio/rucio/blob/master/lib/rucio/client/uploadclient.py#L204

https://github.com/rucio/rucio/blob/master/lib/rucio/rse/rsemanager.py#L350
https://github.com/rucio/rucio/blob/master/lib/rucio/rse/rsemanager.py#L429
https://github.com/rucio/rucio/blob/master/lib/rucio/rse/rsemanager.py#L471

https://github.com/rucio/rucio/blob/master/lib/rucio/rse/protocols/protocol.py#L376
https://github.com/rucio/rucio/blob/master/lib/rucio/rse/protocols/sftp.py#L105

I'm not sure using the REST API is practical for our use case.

Reply:
I don't think tracing the cli code will lead to proper REST API solutions.  Might be better to look at the REST API docs directly:
https://rucio.readthedocs.io/en/latest/rest.html

Reply:
https://rucio.readthedocs.io/en/latest/restapi/did.html#post--dids-(scope)-(name)

Reply:
*sigh* ...

https://github.com/rucio/rucio/blob/master/lib/rucio/core/did.py#L162

And immediately the scene from Kill Bill comes to mind, where Lucy Liu says "You didn't think it was going to be that easy, did you?"

Reply:
Well, going back to the upload case and tracing through the code, it appears this method is responsible for getting a File into Rucio:

https://github.com/rucio/rucio/blob/master/lib/rucio/client/replicaclient.py#L182

This appears to be a `POST /replicas` with the following data:
` data = {'rse': rse, 'files': files, 'ignore_availability': ignore_availability}`

Where files is documented like this:
```
:param files: The list of files. This is a list of DIDs like :
[{'scope': <scope1>, 'name': <name1>}, {'scope': <scope2>, 'name': <name2>}, ...]
```

The dictionary it expects is documented in the `add_replica` method above:
```
        dict = {'scope': scope, 'name': name, 'bytes': bytes, 'meta': meta, 'adler32': adler32}
        if md5:
            dict['md5'] = md5
        if pfn:
            dict['pfn'] = pfn
```

Here `bytes` is the size of the File in bytes, not the actual file data.
Other documentation from the method:
```
        :param scope: The scope of the file.
        :param name: The name of the file.
        :param bytes: The size in bytes.
        :param adler32: adler32 checksum.
        :param pfn: PFN of the file for non deterministic RSE.
        :param md5: md5 checksum.
        :param meta: Metadata attributes.
```